## <center>利用`Cell Ranger`进行单细胞数据分析</center>

#### 主要步骤：
__主要包括四个主要基因表达分析流程__
__`mkfastq`__： 它借鉴了Illumina的bcl2fastq ，可以将一个或多个lane中的混样测序样本按照index标签生成样本对应的fastq文件。即对下机数据base calling files转为fastq文件。
__`count`__：利用mkfastq生成的fq文件，进行比对(基于STAR)、过滤、UMI计数。利用细胞的barcode生成gene-barcode矩阵，然后进行样本分群、基因表达分析。
__`aggr`__：接受cellranger count的输出数据，将同一组的不同测序样本的表达矩阵整合在一起，比如tumor组原来有4个样本，PBMC组有两个样本，现在可以使用aggr生成最后的tumor和PBMC两个矩阵，并且进行标准化去掉测序深度的影响。
__`reanalyze`__：接受cellranger count或cellranger aggr生成的gene-barcode矩阵，使用不同的参数进行降维、聚类。属于定制化分析。

1. 安装cellranger
   需要先注册再安装，参考[使用指南](https://blog.csdn.net/weixin_43784829/article/details/118721346)
   加入环境变量
   ```sh
   export PATH=/Bioinfo/bio_2022_2023_2/bio_nchen/scRNA-analysis/cellranger-7.1.0:$PATH #暂时加入
   ```
<br>

2. 使用`mkref`构建参考基因组
   ```sh
   cellranger mkref --genome=GRCm39 \
                    --nthreads=20 \
                    --fasta=/Bioinfo/bio_2022_2023_2/bio_nchen/Reference/Mus_musculus.GRCm39.dna.primary_assembly.fa \
                    --genes=/Bioinfo/bio_2022_2023_2/bio_nchen/Reference/Mus_musculus.GRCm39.109.chr.gtf
   ```
<br>

3. 使用`count`进行比对
   ```sh
   cellranger count --id=$id \
                    --transcriptome=$ref \
                    --fastqs=$fastqFilePath \
                    --sample=$id \
                    --localcores=15\
                    --nosecondary \
                    --localmem=30
   ```
   使用
   ```sh
   cellranger count \
                    --id=SRR23407934_count \
                    --transcriptome=/Bioinfo/bio_2022_2023_2/bio_nchen/scRNA-analysis/GRCm39 \
                    --sample=SRR23407934 \
                    --fastqs=/Bioinfo/bio_2022_2023_2/bio_nchen/scRNA-analysis/SRR23407934 \
                    --localcores=40\
                    --nosecondary
   ```
   注意：使用是要把fastq.gz文件的命名改成`10x Genomics`的标准命名
---

## <center>利用`Seurat`进行下游分析</center>
1. Read in the packages and data
    ```r
    library(dplyr)
    library(Seurat)
    library(ggplot2)
    library(org.Mm.eg.db)
    library(clusterProfiler)
    rawdata <- Read10X((data.dir = "SRR23407930"), gene.column = 2)
    scObject <- CreateSeuratObject(rawdata)

    # Pre-processing - Transform gene name
    capwords <- function(s, strict = FALSE) {
      cap <- function(s) paste(toupper(substring(s, 1, 1)),
                              {s <- substring(s, 2); if(strict) tolower(s) else s},
                              sep = "", collapse = " " )
      sapply(strsplit(s, split = " "), cap, USE.NAMES = !is.null(names(s)))
    }
    s.genes <- capwords(tolower(cc.genes$s.genes))
    g2m.genes <- capwords(tolower(cc.genes$g2m.genes))
    scObject <- CellCycleScoring(scObject, s.features = s.genes, g2m.features = g2m.genes, set.ident = FALSE)
    ```

2. Pre-processing - wash the data
    ```r
    library(patchwork)
    scObject[["percent.mt"]] <- PercentageFeatureSet(scObject, pattern = "^mt-")
    VlnPlot(scObject, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0.2)
    scObject <- subset(scObject, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 20000)
    plot1 <- FeatureScatter(scObject, feature1 = "nCount_RNA", feature2 = "percent.mt")
    plot2 <- FeatureScatter(scObject, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
    CombinePlots(plots = list(plot1, plot2))
    rm(plot1, plot2)
    ```

3. Pre-processing 
    ```r
    scObject <- NormalizeData(scObject, normalization.method = "RC", scale.factor = 10000)

    scObject <- FindVariableFeatures(scObject, selection.method = "vst", nfeatures = 2000, verbose=FALSE)
    Top5 <- head(VariableFeatures(scObject), 5)
    p1 <- VariableFeaturePlot(scObject)
    LabelPoints(plot = p1, points = Top5, repel = TRUE)  
    rm(p1)

    scObject <- ScaleData(scObject, features = rownames(scObject))

    scObject <- SCTransform(scObject, verbose = TRUE, vars.to.regress = c("percent.mt","nCount_RNA"), variable.features.n = 3000)
    scObject <- RunPCA(scObject, verbose = TRUE, features = VariableFeatures(scObject))
    ElbowPlot(scObject, ndims = 50)
    scObject <- RunUMAP(scObject, dims = 1:8, n.neighbors = 30L, min.dist = 0.01, verbose = FALSE)
    DimPlot(scObject, group.by = "orig.ident", reduction = "umap") +NoLegend()
    DimPlot(scObject, group.by = "Phase", reduction = "umap")
    ```

4. Clustering
    ```r
    scObject <- FindNeighbors(scObject, dims = 1:8, verbose = TRUE)
    scObject <- FindClusters(scObject, verbose = TRUE,resolution = 0.1)
    DimPlot(scObject, group.by = "seurat_clusters",reduction = "umap")

    #Violin plots for phase marker (Mki67) and neuronal precursor marker (Nestin)
    VlnPlot(scObject, features = "Mki67",log = TRUE, pt.size = 0, assay = "RNA") + 
      theme(plot.title = element_text(hjust = 0.5))
    VlnPlot(scObject, features = "Nes", log = TRUE, pt.size = 0, assay = "RNA") + 
      theme(plot.title = element_text(hjust = 0.5))
    #Violin plots for phase marker (Mki67) and neuronal precursor marker (Nestin) grouped by Phase with boxplots
    VlnPlot(scObject, features = "Mki67",log = TRUE,pt.size = 0,group.by = "Phase",assay = "RNA") + 
      theme(plot.title = element_text(hjust = 0.5)) + 
      geom_boxplot(width=0.075,fill="grey")
    VlnPlot(scObject, features = "Nes",log = TRUE,pt.size = 0,group.by = "Phase",assay = "RNA") + 
      theme(plot.title = element_text(hjust = 0.5)) + 
      geom_boxplot(width=0.075,fill="grey")
    ```

5. Find marker
    ```r
    S.marker <- FindMarkers(scObject, assay = "RNA", group.by = "Phase", ident.1 = "S", min.diff.pct = 0.2, verbose = FALSE)
    G1.marker <- FindMarkers(scObject, assay = "RNA", group.by = "Phase", ident.1 = "G1", min.diff.pct = 0.2, verbose = FALSE)
    G2M.marker <- FindMarkers(scObject, assay = "RNA", group.by = "Phase", ident.1 = "G2M", min.diff.pct = 0.2, verbose = FALSE)
    markers <- c("Celf5","Id2","Hmga2","Irx3","Scn1b","Hells", "Pola1", "Mki67")
    DotPlot(scObject, features = markers, group.by = "Phase")
    DotPlot(scObject, features = markers, group.by = "seurat_clusters")
    saveRDS(scObject, "./SRR23407934.rds")
    ```

6. 注释
7. GO富集分析
   构建dataset
   ```r
    scRNA1 <- readRDS("./SRR23407930.rds")
    scRNA2 <- readRDS("./SRR23407934.rds")
    combined <- merge(x = scRNA1, y = c(scRNA2))
    Group <- c(rep("Treatment", dim(scRNA1)[2]), rep("CK", dim(scRNA2)[2]))
    combined <- AddMetaData(combined, Group, col.name = "split")
    DefaultAssay(combined) <- "RNA"
   ```
   标准化数据
   ```r
    combined <- ScaleData(combined, verbose = FALSE)
    combined <- FindVariableFeatures(combined, selection.method = "vst", nfeatures = 3000, verbose = FALSE)
    combined <- RunHarmony(combined, dims = 1:8, "split",verbose = FALSE)
    combined <- RunUMAP(combined,reduction = "harmony", dims = 1:30, verbose = FALSE)
    DimPlot(combined, reduction = "umap", group.by = "split")
    rm(scRNA1, scRNA2)
   ```
   聚类 & 寻找marker
   ```r
    combined <- FindNeighbors(combined, dims = 1:20, reduction = "harmony")
    combined <- FindClusters(combined, resolution = 0.1)
    DimPlot(combined, reduction = "umap", label=TRUE)
    diff_gene <- FindMarkers(combined, ident.1 = "Treatment", ident.2 = "CK", logfc.threshold = 0,min.pct = 0, 
                            test.use="MAST", group.by="split")

    degdf <- subset(diff_gene, p_val_adj<0.05 & abs(avg_log2FC)>0.15)
    degdf$sign <- ifelse(degdf$p_val_adj < 0.0005 & abs(degdf$avg_log2FC) > 1,
                        rownames(degdf), NA)
    P.Value_t = 0.01
    degdf$change = ifelse(degdf$p_val_adj < P.Value_t & degdf$avg_log2FC < -0.5,"down",
                          ifelse(degdf$p_val_adj < P.Value_t & degdf$avg_log2FC > 0.5,"up",
                                "stable"))
    ggplot(degdf, aes(avg_log2FC,  -log10(p_val_adj))) +
      geom_point(alpha=0.4, size=3, aes(color=change)) +
      ylab("-log10(Pvalue)")+
      scale_color_manual(values=c("blue", "grey","red"))+
      geom_hline(yintercept = -log10(P.Value_t),lty=4,col="black",lwd=0.3) +
      geom_vline(xintercept = 0.5,lty=2,col="black",lwd=0.3)+
      geom_vline(xintercept = -0.5,lty=2,col="black",lwd=0.3)+
      geom_label_repel(aes(label=sign), fontface="bold", color="grey50", box.padding=unit(0.35, "lines"), point.padding=unit(0.5, "lines"), segment.colour = "grey50")+
      theme_bw()
   ```
   GO分析
   ```r
    sig_gene <- subset(diff_gene, p_val_adj<0.05 & abs(avg_log2FC)>0.15)
    gene_up=rownames(sig_gene[sig_gene$avg_log2FC > 0.15,])
    gene_down=rownames(sig_gene[sig_gene$avg_log2FC < -0.15,])

    gene_up=as.character(na.omit(AnnotationDbi::select(org.Mm.eg.db,
                                                      keys = gene_up,
                                                      columns = 'ENTREZID',
                                                      keytype = 'SYMBOL')[,2]))
    gene_down=as.character(na.omit(AnnotationDbi::select(org.Mm.eg.db,
                                                        keys = gene_down,
                                                        columns = 'ENTREZID',
                                                        keytype = 'SYMBOL')[,2]))
    ##GO-up
    geneset <- gene_up
    deg_type <- "up"
    ego_CC <- enrichGO(gene = geneset, OrgDb=org.Mm.eg.db, keyType = "ENTREZID",
                      ont = "CC", pAdjustMethod = "BH", minGSSize = 1
                      , pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
    ego_BP <- enrichGO(gene = geneset, OrgDb=org.Mm.eg.db, keyType = "ENTREZID",
                      ont = "BP", pAdjustMethod = "BH", minGSSize = 1,
                      pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
    ego_MF <- enrichGO(gene = geneset, OrgDb=org.Mm.eg.db, keyType = "ENTREZID",
                      ont = "MF", pAdjustMethod = "BH", minGSSize = 1,
                      pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)

    display_number = c(10,10,10)
    ego_result_BP <- as.data.frame(ego_BP)[1:display_number[1], ]
    ego_result_CC <- as.data.frame(ego_CC)[1:display_number[2], ]
    ego_result_MF <- as.data.frame(ego_MF)[1:display_number[3], ]

    go_enrich_df <- data.frame(
      ID=c(ego_result_BP$ID, ego_result_CC$ID, ego_result_MF$ID), Description=c(ego_result_BP$Description,ego_result_CC$Description,ego_result_MF$Description),
      GeneNumber=c(ego_result_BP$Count, ego_result_CC$Count, ego_result_MF$Count),
      type=factor(c(rep("biological process", display_number[1]), 
                    rep("cellular component", display_number[2]),
                    rep("molecular function", display_number[3])), 
                  levels=c("biological process", "cellular component","molecular function" )))

    for(i in 1:nrow(go_enrich_df)){
      description_splite=strsplit(go_enrich_df$Description[i],split = " ")
      description_collapse=paste(description_splite[[1]][1:10],collapse = " ") 
      go_enrich_df$Description[i]=description_collapse
      go_enrich_df$Description=gsub(pattern = "NA","",go_enrich_df$Description)
    }
    go_enrich_df$type_order=factor(rev(as.integer(rownames(go_enrich_df))),labels=rev(go_enrich_df$Description))
    COLS <- c("#66C3A5", "#8DA1CB", "#FD8D62")
    ggplot(data=go_enrich_df, aes(x=type_order,y=GeneNumber, fill=type)) + 
      geom_bar(stat="identity", width=0.8) + 
      scale_fill_manual(values = COLS) + 
      coord_flip() + 
      xlab("GO term") + 
      ylab("Gene_Number") + 
      labs(title = paste0("The Most Enriched GO Terms ",deg_type))+
      theme_bw()
    ## ggsave("./GO-up.png", width = 12, height = 6, dpi = 600)

    ##GO-down
    geneset <- gene_down
    deg_type <- "down"
    ego_CC <- enrichGO(gene = geneset, OrgDb=org.Mm.eg.db, keyType = "ENTREZID",
                      ont = "CC", pAdjustMethod = "BH", minGSSize = 1,
                      pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
    ego_BP <- enrichGO(gene = geneset, OrgDb=org.Mm.eg.db, keyType = "ENTREZID",
                      ont = "BP", pAdjustMethod = "BH", minGSSize = 1,
                      pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
    ego_MF <- enrichGO(gene = geneset, OrgDb=org.Mm.eg.db, keyType = "ENTREZID",
                      ont = "MF", pAdjustMethod = "BH", minGSSize = 1,
                      pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE)
    display_number = c(10,10,10)
    ego_result_BP <- as.data.frame(ego_BP)[1:display_number[1], ]
    ego_result_CC <- as.data.frame(ego_CC)[1:display_number[2], ]
    ego_result_MF <- as.data.frame(ego_MF)[1:display_number[3], ]

    go_enrich_df <- data.frame(
      ID=c(ego_result_BP$ID, ego_result_CC$ID, ego_result_MF$ID), Description=c(ego_result_BP$Description,ego_result_CC$Description,ego_result_MF$Description),
      GeneNumber=c(ego_result_BP$Count, ego_result_CC$Count, ego_result_MF$Count),
      type=factor(c(rep("biological process", display_number[1]), 
                    rep("cellular component", display_number[2]),
                    rep("molecular function", display_number[3])), 
                  levels=c("biological process", "cellular component","molecular function" )))

    for(i in 1:nrow(go_enrich_df)){
      description_splite=strsplit(go_enrich_df$Description[i],split = " ")
      description_collapse=paste(description_splite[[1]][1:10],collapse = " ") 
      go_enrich_df$Description[i]=description_collapse
      go_enrich_df$Description=gsub(pattern = "NA","",go_enrich_df$Description)
    }

    go_enrich_df$type_order=factor(rev(as.integer(rownames(go_enrich_df))),labels=rev(go_enrich_df$Description))
    COLS <- c("#66C3A5", "#8DA1CB", "#FD8D62")
    ggplot(data=go_enrich_df, aes(x=type_order,y=GeneNumber, fill=type)) + 
      geom_bar(stat="identity", width=0.8) + 
      scale_fill_manual(values = COLS) + 
      coord_flip() + 
      xlab("GO term") + 
      ylab("Gene_Number") + 
      labs(title = paste0("The Most Enriched GO Terms ",deg_type))+
      theme_bw()
   ```
8. KEGG富集分析